### PR TITLE
1. Add storage transition to Glacier in storage/glacier.go; 2. Adapt …

### PR DESCRIPTION
**Pull request by brucen1030**
1. Add storage transition to Glacier in storage/glacier.go; 
2. Adapt API with coldstorage; 
3. Change tools/lc.go for transition; 
4. Make necessary changes under meta; 5. Add tools/glacier-util.go for Glacier related test.

Glacier code won't run until "enable_glacier = true" configured in yig.toml.

### CODE REVIEW

**Comment by brucen1030** [comment]
BTW, this commit contains only transition for small file.
Small file offset/length and big file multipart are still under development and will be pushed later.

File: api/bucket-handlers.go [hunk 1]
```diff
@@ -408,6 +409,50 @@ func (api ObjectAPIHandlers) PutBucketLifeCycleHandler(w http.ResponseWriter, r
...
+			// Not supported.
+			helper.Debugln("[", RequestIdFromContext(r.Context()), "]", "StorageClass not supported:", rule, helper.CONFIG.EnableGlacier)
+			WriteErrorResponse(w, r, ErrInvalidStorageClass)
+			return
+		}
+
+		if createVaultFlag == true {
+			err = api.ObjectAPI.CreateVault(r.Context(), credential)
+			if err != nil {
+				helper.Debugln("[", RequestIdFromContext(r.Context()), "]", "create vault failed", bucket)
```
Comment (hopkings2008) [inline api/bucket-handlers.go]: if it is an error, helper.Logger.Error should be used instead of debugln.

File: api/bucket-handlers.go [hunk 2]
```diff
@@ -408,6 +409,50 @@ func (api ObjectAPIHandlers) PutBucketLifeCycleHandler(w http.ResponseWriter, r
...
+				return
+			}
+
+			// Create a hidden bucket for archive restore.
+			err = api.ObjectAPI.MakeBucket(r.Context(),
+				meta.HIDDEN_BUCKET_PREFIX+credential.UserId,
+				Acl{CannedAcl: ValidCannedAcl[CANNEDACL_PRIVATE]},
+				credential)
+			if err != nil && err != ErrBucketAlreadyOwnedByYou {
+				helper.Debugln("[", RequestIdFromContext(r.Context()), "]", "Create hidden bucket failed", err)
```
Comment (hopkings2008) [inline api/bucket-handlers.go]: don't use debugln for error message.

File: api/object-handlers.go [hunk 3]
```diff
@@ -258,6 +258,51 @@ func (api ObjectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
...
+
+	// Get object from hidden bucket.
+	if helper.CONFIG.EnableGlacier && object.StorageClass == meta.ObjectStorageClassGlacier {
+		if object.SseType != "" {
+			// Should no happend. SSE object will not be trasitted as archives.
+			helper.Logger.Println(20, "[", RequestIdFromContext(r.Context()), "]", "GetObjectHandler glacier don't support sse or startoff")
+			WriteErrorResponse(w, r, ErrNotImplemented)
+			return
+		}
+		/*
```
Comment (hopkings2008) [inline api/object-handlers.go]: pls remove the used code.

File: api/object-handlers.go [hunk 4]
```diff
@@ -258,6 +258,51 @@ func (api ObjectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
...
+		if err = api.ObjectAPI.GetObjectFromGlacier(r.Context(), object, version, startOffset, length, writer, credential); err != nil {
+			helper.Logger.Println(20, "[", RequestIdFromContext(r.Context()), "]", "GetObjectHandler GetObjectFromGlacier failed")
+			WriteErrorResponse(w, r, err)
+			return
+		}
+		*/
+
+		// Find restored object name (archive id).
+		restoredObject, err := api.ObjectAPI.GetObjectInfo(r.Context(), 
+														meta.HIDDEN_BUCKET_PREFIX + credential.UserId,
```
Comment (hopkings2008) [inline api/object-handlers.go]: why there are so many whitespaces here? pls use gofmt to format your code.

File: api/object-handlers.go [hunk 5]
```diff
@@ -258,6 +258,51 @@ func (api ObjectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
...
+				RequestIdFromContext(r.Context()), objectName, err)
+			WriteErrorResponse(w, r, ErrInvalidObjectState)
+			return
+		}
+		if err := api.ObjectAPI.GetObject(r.Context(), restoredObject, startOffset, length, writer, sseRequest); err != nil {
+			helper.ErrorIf(err, "Unable to write to client.")
+			if !dataWritten {
+				WriteErrorResponse(w, r, err)
+			}
+			return
```
Comment (hopkings2008) [inline api/object-handlers.go]: the return doesn't response to the client, pls make sure this behavior is correct.

File: api/object-handlers.go [hunk 6]
```diff
@@ -1787,3 +1869,93 @@ func (api ObjectAPIHandlers) PostObjectHandler(w http.ResponseWriter, r *http.Re
...
+	version := r.URL.Query().Get("VersionId") /* TODO */
+
+	// TODO is restore get or put ?
+	credential, err := checkRequestAuth(api, r, policy.GetObjectAction, bucketName, objectName)
+	if err != nil {
+		WriteErrorResponse(w, r, err)
+		return
+	}
+
+	buffer, err := ioutil.ReadAll(io.LimitReader(r.Body, 1024))
```
Comment (hopkings2008) [inline api/object-handlers.go]: 1024 means that the size of request body is fixed 1024?

File: coldstorage/client/glacierclient/vault.go [hunk 7]
```diff
@@ -4,7 +4,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/glacier"
-	. "github.com/journeymidnight/yig/coldstorage/client"
+	//. "github.com/journeymidnight/yig/coldstorage/client"
```
Comment (sunfch) [inline coldstorage/client/glacierclient/vault.go]: pls delete unused code

File: conf/yig.toml [hunk 8]
```diff
@@ -52,6 +52,12 @@ cache_circuit_open_threshold = 1
 # Ceph Config
 ceph_config_pattern = "/etc/ceph/*.conf"
 
+# Glacier Config
+enable_glacier = false
+glacier_host = ""
+glacier_region = "us-east-1"
+glacier_tier = "Standard"
+
```
Comment (sunfch) [inline conf/yig.toml]: Pls place the items about glacier in a group

File: api/bucket-handlers.go (hunk 1 above)
Comment (brucen1030) [reply api/bucket-handlers.go]: OK. I'll use helper.Logger.Println(5 instead.

File: api/bucket-handlers.go (hunk 2 above)
Comment (brucen1030) [reply api/bucket-handlers.go]: OK.

File: api/object-handlers.go (hunk 3 above)
Comment (brucen1030) [reply api/object-handlers.go]: OK.

File: api/object-handlers.go (hunk 4 above)
Comment (brucen1030) [reply api/object-handlers.go]: OK.

File: api/object-handlers.go (hunk 5 above)
Comment (brucen1030) [reply api/object-handlers.go]: Yes. Just copied from common S3 object GET handling several lines after it.
Please find the comments.
	// Reads the object at startOffset and writes to mw.
	if err := api.ObjectAPI.GetObject(r.Context(), object, startOffset, length, writer, sseRequest); err != nil {
		helper.ErrorIf(err, "Unable to write to client.")
		if !dataWritten {
			// Error response only if no data has been written to client yet. i.e if
			// partial data has already been written before an error
			// occurred then no point in setting StatusCode and
			// sending error XML.
			WriteErrorResponse(w, r, err)
		}
		return
	}

File: api/object-handlers.go (hunk 6 above)
Comment (brucen1030) [reply api/object-handlers.go]: It should be less 1024.
The API is defined as follows.
And I need only the first several params excluding "SelectParameters" and "OutputLocation". "SelectParameters" and "OutputLocation" are not supported in this release.
1024 long should be enough from "Days" to "Tier".

https://docs.aws.amazon.com/AmazonS3/latest/API/API_RestoreObject.html
POST /{Key+}?restore&VersionId=VersionId HTTP/1.1
Host: Bucket.s3.amazonaws.com
x-amz-request-payer: RequestPayer
<?xml version="1.0" encoding="UTF-8"?>
<RestoreRequest xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
   <Days>integer</Days>
   <GlacierJobParameters>
      <Tier>string</Tier>
   </GlacierJobParameters>
   <Type>string</Type>
   <Tier>string</Tier>
   <Description>string</Description>
   <SelectParameters>
      <Expression>string</Expression>
      <ExpressionType>string</ExpressionType>
      <InputSerialization>
         <CompressionType>string</CompressionType>
         <CSV>
            <AllowQuotedRecordDelimiter>boolean</AllowQuotedRecordDelimiter>
            <Comments>string</Comments>
            <FieldDelimiter>string</FieldDelimiter>
            <FileHeaderInfo>string</FileHeaderInfo>
            <QuoteCharacter>string</QuoteCharacter>
            <QuoteEscapeCharacter>string</QuoteEscapeCharacter>
            <RecordDelimiter>string</RecordDelimiter>
         </CSV>
         <JSON>
            <Type>string</Type>
         </JSON>
         <Parquet>
         </Parquet>
      </InputSerialization>
      <OutputSerialization>
         <CSV>
            <FieldDelimiter>string</FieldDelimiter>
            <QuoteCharacter>string</QuoteCharacter>
            <QuoteEscapeCharacter>string</QuoteEscapeCharacter>
            <QuoteFields>string</QuoteFields>
            <RecordDelimiter>string</RecordDelimiter>
         </CSV>
         <JSON>
            <RecordDelimiter>string</RecordDelimiter>
         </JSON>
      </OutputSerialization>
   </SelectParameters>
   <OutputLocation>
      <S3>
         <AccessControlList>
            <Grant>
               <Grantee>
                  <DisplayName>string</DisplayName>
                  <EmailAddress>string</EmailAddress>
                  <ID>string</ID>
                  <xsi:type>string</xsi:type>
                  <URI>string</URI>
               </Grantee>
               <Permission>string</Permission>
            </Grant>
         </AccessControlList>
         <BucketName>string</BucketName>
         <CannedACL>string</CannedACL>
         <Encryption>
            <EncryptionType>string</EncryptionType>
            <KMSContext>string</KMSContext>
            <KMSKeyId>string</KMSKeyId>
         </Encryption>
         <Prefix>string</Prefix>
         <StorageClass>string</StorageClass>
         <Tagging>
            <TagSet>
               <Tag>
                  <Key>string</Key>
                  <Value>string</Value>
               </Tag>
            </TagSet>
         </Tagging>
         <UserMetadata>
            <MetadataEntry>
               <Name>string</Name>
               <Value>string</Value>
            </MetadataEntry>
         </UserMetadata>
      </S3>
   </OutputLocation>
</RestoreRequest>

File: coldstorage/client/glacierclient/vault.go (hunk 7 above)
Comment (brucen1030) [reply coldstorage/client/glacierclient/vault.go]: OK.

File: conf/yig.toml (hunk 8 above)
Comment (brucen1030) [reply conf/yig.toml]: OK.

File: api/datatype/lifecycle.go [hunk 9]
```diff
@@ -3,17 +3,41 @@ package datatype
 import (
 	"encoding/xml"
 	//	. "github.com/journeymidnight/yig/error"
```
Comment (sunfch) [inline api/datatype/lifecycle.go]: pls delete unused import line

File: api/object-handlers.go [hunk 10]
```diff
@@ -258,6 +258,51 @@ func (api ObjectAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Req
 			r.Header.Get("X-Amz-Server-Side-Encryption-Customer-Key-Md5"))
 	}
 
+	helper.Logger.Println(20, "[", RequestIdFromContext(r.Context()), "]", "GetObjectHandler", object.StorageClass, meta.ObjectStorageClassGlacier, object.VersionId, version)
+
+	// Get object from hidden bucket.
+	if helper.CONFIG.EnableGlacier && object.StorageClass == meta.ObjectStorageClassGlacier {
```
Comment (sunfch) [inline api/object-handlers.go]: should we handle glacier object firstly?

File: api/datatype/lifecycle.go (hunk 9 above)
Comment (brucen1030) [reply api/datatype/lifecycle.go]: ok.

File: api/object-handlers.go (hunk 10 above)
Comment (brucen1030) [reply api/object-handlers.go]: An object will be in either Standard or Glacier.
Here we check if it's in Glacier, get the restored object instead, and return in line 296.

**Pull request merged**